bindings/generate_bindings.py

```diff
@@ -0,0 +1,600 @@
+        return out
+
+    def gen_def(self) -> Code:
+        # Function name.
+        out = [f"def {self.method}_{self.name}("]
+
+        # Function parameters.
+        out += ['    do_request: Request,']
+        if self.params:
+            out += ["    *,"]
+
+        required = sorted(p for p in self.params if self.params[p].required)
+        optional = sorted(p for p in self.params if not self.params[p].required)
+        for name in required + optional:
+            out += [self.params[name].gen_function_param()]
+
+        # Function return type.
+        # (simplifying assumptions; if broken we need more logic)
+        responses = {**self.responses}
+        default = responses.pop("default")
+        assert isinstance(default, Ref) and default.name == "runtimeError", (self.name, default)
+
+        if len(responses) == 1:
+            returntype = next(iter(responses.values()))
+            returntypestr = returntype.annotation()
+        else:
+            returntypes = set(r.annotation(prequoted=True) for r in responses.values())
+            returntypestr = '"Union[' + ", ".join(sorted(returntypes)) + ']"'
+        assert len(responses) == 1, (self.name, responses)
+
+        out += [f') -> {returntypestr}:']
+
+        # Function body.
+        path_params = sorted(p for p in self.params if self.params[p].where == "path")
+        body_params = sorted(p for p in self.params if self.params[p].where == "body")
+        query_params = sorted(p for p in self.params if self.params[p].where == "query")
+
+        pathstr = f'"{self.path}"'
+        if path_params:
+            # Happily, we can just generate an f-string based on the path swagger gives us.
+            pathstr = "f" + pathstr
+
+        if query_params:
+            out += ["    _params = {"]
+            for p in query_params:
+                out += [f'        "{self.params[p].serialized_name}": {self.params[p].name},']
+            out += ["    }"]
+        else:
+            out += ["    _params = None"]
+
+        if "body" in self.params:
+            bodystr = self.params["body"].dump()
+        else:
+            bodystr = "None"
+        out += ["    _resp = do_request("]
+        out += [f"        \"{self.method.upper()}\","]
+        out += [f"        {pathstr},"]
+        out += ["        _params,"]
+        out += [f"        {bodystr},"]
+        out += ["    )"]
+        for expect, returntype in responses.items():
+            out += [f"    if _resp.status_code == {expect}:"]
+            if returntype.isnone():
+                out += ['        return']
+            else:
+                out += [f'        return {returntype.load("_resp.json()")}']
+        out += [f'    raise APIHttpError("{self.method}_{self.name}", _resp)']
+
+        return "\n".join(out)
+
+
+def classify_type(path: str, schema: dict) -> TypeAnno:
+    # enforce valid jsonschema:
+    assert isinstance(schema, dict), (path, schema)
+    # XXX: how to handle inlined enums?
```

Comment:
this is talking about the following situation where some parameters are defined inline as the following:

```
// params:...
            {
              "name": "orderBy",
              "description": "Order commands in either ascending or descending order.\n\n - ORDER_BY_UNSPECIFIED: Returns records in no specific order.\n - ORD
  ER_BY_ASC: Returns records in ascending order.\n - ORDER_BY_DESC: Returns records in descending order.",
              "in": "query",
              "required": false,
              "type": "string",
*             "enum": [
                "ORDER_BY_UNSPECIFIED",
                "ORDER_BY_ASC",
                "ORDER_BY_DESC"
              ],
              "default": "ORDER_BY_UNSPECIFIED"
            },
```
instead of being a reference to the existing enum:

```
// definitions..
*     "v1OrderBy": {
        "type": "string",
        "enum": [
          "ORDER_BY_UNSPECIFIED",
          "ORDER_BY_ASC",
          "ORDER_BY_DESC"
        ],
        "default": "ORDER_BY_UNSPECIFIED",
        "description": "Order records in either ascending or descending order.\n\n - ORDER_BY_UNSPECIFIED: Returns records in no specific order.\n - ORDER_BY_A  SC: Returns records in ascending order.\n - ORDER_BY_DESC: Returns records in descending order.",
*       "title": "OrderBy"
      },
```

Not sure if this is a result of how we generated the openapi spec or if this is standard. there is a mention of reusable enums in OpenAPI3
which might be what we're looking for
https://swagger.io/docs/specification/data-models/enums/

Reply:
without direct a reference from the swagger spec, we could attempt to link them ourselves. what else could we do? 🤔

Reply:
Yeah I like how you did it there.

Reply:
added a todo for the streaming portion. and addressed the one about asserting no `.` outside curly braces in `path`